nativescript-3.2.0-memory-management/README.md

```diff
@@ -0,0 +1,126 @@
+
+### Properties of the Implementation
+#### Native Instances can Easily Leak
+You can make a reference cycle in Objective-C that will leak native and JavaScript instances, because the JavaScript GC does not traverse the native objects and fields to find cycles. Native tools can point to leaked instances.
+
+#### Native Instances can be Prematurely Collected
+You can let instances be prematurely collected. Using weak properties or API, that involve methods such as `setTarget:selector:...` that will add the Objective-C instance as native target, but holds weak Objective-C references, these does not increment the ref-count of the Objective-C instance. The target may remain with ref-count 1. If the JavaScript GC collects the JavaScript instances of the splice, it will then delete the Objective-C instance. The annoying part is the code will behave properly for a while - until the non deterministic completion of the GC.
+
+#### Half-Dead Splice
+We haven't had problems with cases in which the JavaScript instance be alive and the Objective-C instance be deallocated or the vice versa.
+
+#### Very Objective-C Friendly
+Overall the implementation is really Objective-C friendly and predictable. When working with native APIs it requires some additional care about memory management, but nothing more than general iOS knowledge. Is very UI friendly and does not introduce big pauses on the main UI thread.
+
+## Android
+In Android both the Java and the JavaScript VMs are GC based. The Android VM has limited public API to subscribe for GC events, while in the V8 there is an API to subscribe for GC prologue and epilogue. Also V8 allows us to subscribe for notifications when a JavaScript instance is marked for collection and can be optionally revived.
+
+### Splice Life-Cycle
+The Android splice has two flavors:
+ - It is considered to "have implementation object" in cases when:
+    - A splice is created for "anonymous interface", such as `new ClickListener({ ... })`.
+    - A splice is creates for "extended native class", such as `var MyView = View.extends({ ... }); var myView = new MyView();`.
+ - It is considered **not to** "have implementation object" in cases when:
+    - A splice is created for `var button = new android.widget.Button(...)` where a native class is instantiated.
+    - A splice is created for `var i = anAndroidObject.getValue()` when Java instance is returned by the getValue() method.
+
+When a splice is created
+ - The splice have strong reference to the Java instance, and the Java instance can not be collected by the Android VM GC
+
+On V8 GC collection phase:
+ - For each splice that has implementation object, from its JavaScript instance, all other reachable JavaScript instances are traversed. For these reachable JavaScript objects:
+    - If the reached object is an implementation object - the splice makes the reference to the Java instance weak.
```

Comment:
I think this is wrong.

Here if from a weak JavaScript instance of a splice that has implementation object, we reach another splice that has implementation object, we clean the second from m_implObjStrong: https://github.com/NativeScript/android-runtime/blob/master/runtime/src/main/jni/ObjectManager.cpp#L433

Next it won't be in the m_implObjStrong list here:
https://github.com/NativeScript/android-runtime/blob/master/runtime/src/main/jni/ObjectManager.cpp#L612

And if the splice had strong reference to the Java instance, it will remain strong here:
https://github.com/NativeScript/android-runtime/blob/master/runtime/src/main/jni/ObjectManager.cpp#L668

So the statement here, should be that the reached splice will be marked, so it won't make the reference to its Java instance weak in the next step.